src/Manager.php

```diff
@@ -0,0 +1,703 @@
+
+use Yiisoft\Access\AccessCheckerInterface;
+use InvalidArgumentException;
+use RuntimeException;
+
+/**
+ * Manager represents an authorization manager that stores authorization
+ * For more details and usage information on Manager, see the [guide article on security authorization](guide:security-authorization).
+ */
+final class Manager implements AccessCheckerInterface
+{
+    private Storage $storage;
+    private RuleFactoryInterface $ruleFactory;
+
+    /**
+     * @var array a list of role names that are assigned to every user automatically without calling {@see assign()}.
+     * Note that these roles are applied to users, regardless of their state of authentication.
+     */
+    private array $defaultRoles = [];
+
+    public function __construct(Storage $storage, RuleFactoryInterface $ruleFactory)
+    {
+        $this->storage = $storage;
+        $this->ruleFactory = $ruleFactory;
+    }
+
+    /**
+     * @param mixed $userId
+     * @param string $permissionName
+     * @param array $parameters
+     * @return bool
+     * @throws RuntimeException
+     */
+    public function userHasPermission($userId, string $permissionName, array $parameters = []): bool
+    {
+        $assignments = $this->storage->getUserAssignments($userId);
+
+        if (empty($assignments) && $this->defaultRolesIsEmpty()) {
+            return false;
+        }
+
+        if ($this->storage->getPermissionByName($permissionName) === null) {
+            return false;
+        }
+
+        return $this->userHasPermissionRecursive($userId, $permissionName, $parameters, $assignments);
+    }
+
+    /**
+     * Checks the possibility of adding a child to parent.
+     *
+     * @param Item $parent the parent item
+     * @param Item $child the child item to be added to the hierarchy
+     *
+     * @return bool possibility of adding
+     */
+    public function canAddChild(Item $parent, Item $child): bool
+    {
+        return $parent->canBeParentOfItem($child) && !$this->detectLoop($parent, $child);
+    }
+
+    /**
+     * Adds an item as a child of another item.
+     *
+     * @param Item $parent
+     * @param Item $child
+     *
+     * @throws RuntimeException
+     * @throws InvalidArgumentException
+     */
+    public function addChild(Item $parent, Item $child): void
+    {
+        if (!$this->hasItem($parent->getName()) || !$this->hasItem($child->getName())) {
+            throw new InvalidArgumentException(
+                sprintf('Either "%s" or "%s" does not exist.', $parent->getName(), $child->getName())
+            );
+        }
+
+        if ($parent->isEqualName($child->getName())) {
+            throw new InvalidArgumentException(sprintf('Cannot add "%s" as a child of itself.', $parent->getName()));
+        }
+
+        if (!$parent->canBeParentOfItem($child)) {
+            throw new InvalidArgumentException(
+                sprintf('Can not add "%s" role as a child of "%s" permission.', $child->getName(), $parent->getName())
```

Comment:
Here we are assuming that `canBeParentOfItem` may fail for a single reason. If so, it worth moving the check here and deleting `Item::canBeParentOfItem()`.

Reply:
fixed